components/script/dom/htmlanchorelement.rs

```diff
@@ -574,45 +574,59 @@ impl Activatable for HTMLAnchorElement {
     }
 }
 
-/// <https://html.spec.whatwg.org/multipage/#the-rules-for-choosing-a-browsing-context-given-a-browsing-context-name>
-fn is_current_browsing_context(target: DOMString) -> bool {
-    target.is_empty() || target == "_self"
-}
-
 /// <https://html.spec.whatwg.org/multipage/#following-hyperlinks-2>
 pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>, referrer_policy: Option<ReferrerPolicy>) {
-    // Step 1: replace.
-    // Step 2: source browsing context.
-    // Step 3: target browsing context.
-    let target = subject.get_attribute(&ns!(), &local_name!("target"));
+    // Step 1: TODO: If subject cannot navigate, then return.
+    // Step 2.
+    let replace = false;
+
+    let document = document_from_node(subject);
+    let window = document.window();
+
+    // Step 3: source browsing context.
+    let source = document.browsing_context().unwrap();
 
-    // Step 4: disown target's opener if needed.
+    // Step 4-5: target attribute.
+    let target_attribute_value = subject.get_attribute(&ns!(), &local_name!("target"));
+
+    // Step 6.
+    let noopener = if let Some(link_types) = subject.get_attribute(&ns!(), &local_name!("rel")) {
+        let values = link_types.Value();
+        let contains_noopener = values.contains("noopener");
+        let contains_noreferrer = values.contains("noreferrer");
+        contains_noreferrer || contains_noopener
+    } else {
+        false
+    };
+
+    // Step 7.
+    let (maybe_chosen, _new) = match target_attribute_value {
+          Some(name) => source.choose_browsing_context(name.Value(), noopener),
+          None => (Some(window.window_proxy()), false)
+    };
+    let chosen = match maybe_chosen {
+        Some(proxy) => proxy,
+        None => return,
+    };
+    let target_document = chosen.document().unwrap();
+    let target_window = target_document.window();
+
+    // Step 9, TODO: disown target's opener if needed.
```

Comment:
Currently disowning in done inside `create_auxiliary_browsing_context`, and for better spec compliance I should probably move it out into the various steps of the algorithms using the auxiliary as target(noopener is only relevant as part of creating the browsing context in order to determine whether the local storage should be copied or not)